File: java/src/demos/dlineage/dataflow/model/AbstractRelation.java

package demos.dlineage.dataflow.model;

import java.util.ArrayList;
import java.util.List;

public abstract class AbstractRelation implements Relation
{

	public static int RELATION_ID = 0;

	private int id;

	protected RelationElement<?> target;
	protected List<RelationElement<?>> sources = new ArrayList<RelationElement<?>>( );

	public AbstractRelation( )
	{
		id = ++RELATION_ID;
	}

	public int getId( )
	{
		return id;
	}

	public RelationElement<?> getTarget( )
	{
		return target;
	}

	public void setTarget( RelationElement<?> target )
	{
		this.target = target;
	}

	public RelationElement<?>[] getSources( )
	{
		return sources.toArray( new RelationElement<?>[0] );
	}

	public void addSource( RelationElement<?> source )
	{
		if ( source != null && !sources.contains( source ) )
		{
			sources.add( source );
		}
	}
}

File: dotnet/src/demos/dataFlowAnalyzer/dataflow/model/AbstractRelation.cs
using System.Collections.Generic;

namespace gudusoft.gsqlparser.demos.dlineage.dataflow.model
{


	public abstract class AbstractRelation : Relation
	{
		public abstract RelationType RelationType {get;}

		public static int RELATION_ID = 0;

		private int id;

		protected internal RelationElement target;

        protected internal List<RelationElement> sources = new List<RelationElement>();

		public AbstractRelation()
		{
			id = ++RELATION_ID;
		}

		public virtual int Id
		{
			get
			{
				return id;
			}
		}

		public virtual RelationElement Target
		{
			get
			{
				return target;
			}
			set
			{
				this.target = value;
			}
		}


		public virtual RelationElement[] Sources
		{
			get
			{
				return sources.ToArray();
			}
		}

		public virtual void addSource(RelationElement source)
		{
			if (source != null && !sources.Contains(source))
			{
				sources.Add(source);
			}
		}
	}

}